Mathlib/RingTheory/Spectrum/Prime/ConstructibleSet.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2024 Yaël Dillies, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Andrew Yang
+-/
+import Mathlib.Order.SuccPred.WithBot
+import Mathlib.RingTheory.Spectrum.Prime.Topology
+
+/-!
+# Constructible sets in the prime spectrum
+
+This file provides tooling for manipulating constructible sets in the prime spectrum of a ring.
+
+## TODO
+
+Link to constructible sets in a topological space.
+-/
+
+open Finset
+open scoped Polynomial
+
+namespace PrimeSpectrum
+variable {R S T : Type*} [CommRing R] [CommRing S] [CommRing T]
+
+variable (R) in
+/-- The data of a constructible set `s` is finitely many tuples `(f, g₁, ..., gₙ)` such that
+`s = ⋃ (f, g₁, ..., gₙ), V(g₁, ..., gₙ) \ V(f)`.
+
+To obtain `s` from its data, use `PrimeSpectrum.ConstructibleSetData.toSet`. -/
+abbrev ConstructibleSetData := Finset (Σ n, R × (Fin n → R))
```

Comment:
The mixture of `Finset` and `Fin`-tuple in this definition is an interesting desing choice. Can you explain it (in connection to your use case) in the docstring? Why not use `Finset (R × Σ n, Fin n → R)`? How about `Finset (R × Multiset R)` or `Finset (R ×Finset R)`? (I guess the sum in `degBound` won't behave well under `Finset.map` but `Multiset.map` should be fine?) How about replacing the outer `Finset` by `Multiset` to avoid `DecidableEq` required by `Finset.map`?

Reply:
`Finset (R × Σ n, Fin n → R)` is different from what we have here. It only has one $$f \in R$$ but we have a $$f_i$$ and a family of $$g_{i, j}$$ for each component $$i$$. And we use indexed family because we need a fixed number of $$g_{i, j}$$'s for each $$i$$. Changing the outermost `Finset` to `Multiset` might be fine but I'm not sure if it is an improvement.

Reply:
I'm not sure I understand; they are in canonical bijection because products distribute over sums:
```
import Mathlib
example : (Σ n, R × (Fin n → R)) ≃ R × Σ n, Fin n → R where
  toFun f := ⟨f.2.1, f.1, f.2.2⟩
  invFun f := ⟨f.2.1, f.1, f.2.2⟩
  left_inv _ := rfl
  right_inv _ := rfl
```
I seems you are giving a different reason for the choice of `Fin`-tuple over Finset than the one I guessed; it would be helpful to explain in the docstring what you mean by "a fixed number of $g_{i,j}$'s", and where and how you need them.

Reply:
The reason we want to deal with indexed families rather than (fin)sets is that indexed families are easier to deal with than (fin)sets. I'm sure we could Chevalley with (fin)sets but it doesn't really seem worth trying

Reply:
That's fair, thanks.
maintainer delegate